File: python/menrestroom/menrestroom.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""Men Restroom Algorithm"""

from __future__ import (division, absolute_import, print_function,
                        unicode_literals)

import random
import time
import threading


class Menpeeing:
    def __init__(self,
                 stalls=10,
                 stallfreq=2,
                 mintimepeeing=1,
                 maxtimepeeing=10):
        self.stallfreq = stallfreq
        self.mintimepeeing = mintimepeeing
        self.maxtimepeeing = maxtimepeeing
        self.timepeeing = random.randint(self.mintimepeeing,
                                         self.maxtimepeeing)
        self.untaken = list(range(0, stalls))
        self.taken = []
        self.new_stall = round(sum(self.untaken) / len(self.untaken) + .5)
        self.left = self.untaken[1:self.new_stall:2]
        self.right = self.untaken[self.new_stall + 2::2]
        self.emo_empty = "\U0001F6BD"
        self.emo_taken = "\U0001F6B6"
        self.emo_door = "\U0001F6AA"
        self.stall_print = list(self.emo_empty * stalls + self.emo_door)

        self.take_stall()
        self.leave_stall()

    def take_stall(self):
        # Check if there is any empty stall
        if self.untaken:
            if not self.taken:
                new_stall = self.new_stall
            else:
                if self.left:
                    new_stall = random.choice(self.left)
                    self.left.remove(new_stall)
                elif self.right:
                    new_stall = random.choice(self.right)
                    self.right.remove(new_stall)
                else:
                    new_stall = random.choice(self.untaken)

            self.untaken.remove(new_stall)
            self.taken.append(new_stall)

        self.stall_print.pop(self.taken[-1])
        self.stall_print.insert(self.taken[-1], self.emo_taken)

        threading.Timer(self.stallfreq, self.take_stall).start()

        return self.untaken, self.taken, self.stall_print

    def leave_stall(self):
        # Pee done
        if self.taken:
            old_stall = self.taken[0]
            self.taken.remove(old_stall)
            self.untaken.append(old_stall)
            self.stall_print.pop(old_stall)
            self.stall_print.insert(old_stall, self.emo_empty)
            self.timepeeing = random.randint(self.mintimepeeing,
                                             self.maxtimepeeing)

        threading.Timer(self.timepeeing, self.leave_stall).start()

        return self.untaken, self.taken, self.stall_print


def main():

    newpee = Menpeeing(stalls=10, stallfreq=1)

    while newpee.untaken:
        for item in newpee.stall_print:
            print(item, end=' ', flush=True)

        time.sleep(newpee.stallfreq)
        print('\n')


if __name__ == '__main__':
    main()

File: go/menrestroom/menrestroom.go
///bin/true; exec /usr/bin/env go run "" "$@"
package main

import (
	"fmt"
	"math"
	"math/rand"
	"sort"
	"strings"
	"time"
)

var untaken []int
var taken []int
var timePeeing time.Duration
var newStall int
var left []int
var right []int
var stall int
var stallPrint []string

const stalls int = 10
const stallFreq time.Duration = 1
const mintimepeeing int = 1
const maxtimepeeing int = 10
const emoEmpty string = "\U0001F6BD"
const emoTaken string = "\U0001F6B6"
const emoDoor string = "\U0001F6AA"
const halfValue float32 = 0.5

func makeRange(min, max int) []int {
	a := make([]int, max-min+1)
	for i := range a {
		a[i] = min + i
	}
	return a
}

func sumArray(array []int) int {
	result := 0
	for _, numb := range array {
		result += numb
	}
	return result
}

func sliceArray(array []int, start int) []int {

	result := []int{}
	for i := start; i < len(array); i += 2 {
		result = append(result, array[i])
	}
	return result
}

func index(slice []int, item int) int {
	for result := range slice {
		if slice[result] == item {
			return result
		}
	}
	return -1
}

func init() {

	rand.Seed(time.Now().UnixNano())
	timePeeing = time.Duration(rand.Intn(maxtimepeeing-mintimepeeing+1) + mintimepeeing)
	untaken = makeRange(1, stalls)
	newStall = int(math.Floor(float64(sumArray(untaken)) / float64(len(untaken))))
	if stalls%2 == 0 {
		left = sliceArray(untaken[0:newStall-1], 0)
	} else {
		left = sliceArray(untaken[0:newStall-1], 1)
	}
	right = sliceArray(untaken[newStall:], 1)
	stallPrint = strings.SplitN(strings.Repeat(emoEmpty, stalls)+emoDoor, "", stalls+1)

}

func takeStall() []string {

	if len(untaken) > 0 {
		if len(taken) == 0 {
			stall = newStall
		} else {
			if len(left) > 0 {
				randomIndex := rand.Intn(len(left))
				stall = left[randomIndex]
				left = append(left[:randomIndex], left[randomIndex+1:]...)
			} else if len(right) > 0 {
				randomIndex := rand.Intn(len(right))
				stall = right[randomIndex]
				right = append(right[:randomIndex], right[randomIndex+1:]...)

			} else {
				randomIndex := rand.Intn(len(untaken))
				stall = untaken[randomIndex]
			}
		}
		stallIndex := index(untaken, stall)
		untaken = append(untaken[:stallIndex], untaken[stallIndex+1:]...)
	}

	taken = append(taken, stall)

	stallPrint[taken[len(taken)-1]-1] = emoTaken

	// leaveTimer := time.NewTimer(timePeeing)
	// <-leaveTimer.C
	// time.Sleep(timePeeing)
	leaveTimer := time.NewTimer(timePeeing * time.Second)

	select {
	case <-leaveTimer.C:
		leaveStall()
	}

	return stallPrint
}

func leaveStall() {

	if len(taken) > 0 {
		oldStall := taken[0]
		taken = append(taken[:0], taken[1:]...)
		untaken = append(untaken, oldStall)
		sort.Ints(untaken)
		stallPrint[oldStall-1] = emoEmpty
		timePeeing = time.Duration(rand.Intn(maxtimepeeing-mintimepeeing+1) + mintimepeeing)
	}

	return
}

func main() {

	//takeTicker := time.NewTicker(stallFreq * time.Second)
	//leaveTicker := time.NewTicker(timePeeing)

	for len(untaken) > 0 {
		fmt.Println("\033[H\033[2J")

		for _, item := range stallPrint {
			fmt.Print(item + " ")

		}

		takeTimer := time.NewTimer(stallFreq * time.Second)

		select {
		case <-takeTimer.C:
			go takeStall()
		}

		//time.Sleep(stallFreq * time.Second)

		//go takeStall()
		//go leaveStall()

	}

	fmt.Println("\033[H\033[2J")
	for _, item := range stallPrint {
		fmt.Print(item + " ")
	}

	fmt.Println()

}
